### Take Test: WEEK 5: FRIDAY LAB 04
 

In [19]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
from scipy import spatial
import statsmodels.api as sm
from statsmodels.formula.api import ols

from sklearn import linear_model
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

#Reading in the Olympic history: athletes and results  as olympic and results dataframe
olympic = pd.read_csv('archive/athlete_events.csv')


#Verifying that we can see the data
olympic.head()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

In [20]:
olympic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 20.7+ MB


1. Select one dependent variable in the dataset. This will be your target variable for the remainder of this lab. Explain how and why you have selected that particular target variable in no less than 3 sentences.

The Chosen dependant value is the medal. Medal represent the various achievements of the athletes on the various sports through the various sports. It allows for the prediction of the Medal types.

2. Determine what correlations there are among independent variables (= predictors) in the dataset by setting up a correlation table.  Explain in at least 2 sentences what correlations your analysis is showing and if any predictor variables appear to be correlated to one another

In [21]:
olympic.corr()

ID       Age    Height    Weight      Year
ID      1.000000 -0.003631 -0.011141 -0.009176  0.011885
Age    -0.003631  1.000000  0.138246  0.212069 -0.115137
Height -0.011141  0.138246  1.000000  0.796213  0.047578
Weight -0.009176  0.212069  0.796213  1.000000  0.019095
Year    0.011885 -0.115137  0.047578  0.019095  1.000000

Height and weight have a moderate positive correlation. This shows that heavier athletes are also likely to be tall also.

3. Determine which independent variable in the dataset is the most important overall predictor for your chosen dependent (target) variable, then build a simple linear regression based on this predictor and evaluate the quality of your model. Explain why (or why not) it is acceptable to use this model to predict future values of your dependent variable.

The most important independent variable is the Age as athletes perform better and win most medals in their younger years

In [22]:
olympic = olympic[olympic['Age'].notna()]
olympic['Medal'] = pd.factorize(olympic['Medal'])[0]
X = olympic['Age']
y = olympic['Medal']
X = sm.add_constant(X)
olympic.mod1 = sm.OLS(y, X).fit()  ## sm.OLS(output, i.e. dependent variable, input, i.e. independent variable)
olympic.mod1_summary = olympic.mod1.summary()
print(olympic.mod1_summary)

                            OLS Regression Results                            
Dep. Variable:                  Medal   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     138.4
Date:                Fri, 11 Feb 2022   Prob (F-statistic):           5.99e-32
Time:                        17:12:52   Log-Likelihood:            -3.0484e+05
No. Observations:              261642   AIC:                         6.097e+05
Df Residuals:                  261640   BIC:                         6.097e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.7744      0.006   -123.906      0.0

The p-value is 0.000 which shows that both chosen variables are dependant.  It is acceptable to use this model to predict future values of Medal awards, this is because R squared is at 0.1%.

4. Run a basic multiple regression and review the p-values for each variable. If any variables have p-values > 0.05, remove them because they are not statistically significant. Rerun the multiple regression until you can see which of the variables in your dataset should be combined to produce an optimal model. Then build and run your multiple regression model and test its quality. Explain how this model improves the output of the simple linear regression you ran previously.

In [23]:
olympic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 261642 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      261642 non-null  int64  
 1   Name    261642 non-null  object 
 2   Sex     261642 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210068 non-null  float64
 5   Weight  207379 non-null  float64
 6   Team    261642 non-null  object 
 7   NOC     261642 non-null  object 
 8   Games   261642 non-null  object 
 9   Year    261642 non-null  int64  
 10  Season  261642 non-null  object 
 11  City    261642 non-null  object 
 12  Sport   261642 non-null  object 
 13  Event   261642 non-null  object 
 14  Medal   261642 non-null  int32  
dtypes: float64(3), int32(1), int64(2), object(9)
memory usage: 22.0+ MB


In [27]:
# storing only the factor levels array inplace of the object utilizing pandas factorize to utilize all the dataset variables 

olympic['Sex'] = pd.factorize(olympic['Sex'])[0]
olympic['Name'] = pd.factorize(olympic['Name'])[0]
olympic['Team'] = pd.factorize(olympic['Team'])[0]
olympic['NOC'] = pd.factorize(olympic['NOC'])[0]
olympic['Games'] = pd.factorize(olympic['Games'])[0]
olympic['Season'] = pd.factorize(olympic['Season'])[0]
olympic['City'] = pd.factorize(olympic['City'])[0]
olympic['Sport'] = pd.factorize(olympic['Sport'])[0]
olympic['Event'] = pd.factorize(olympic['Event'])[0]

# multiple regression
X = olympic.iloc[:, :-1].values
y = olympic.iloc[:, 14].values
X = np.nan_to_num(X)

X = np.append(arr = np.ones((261642, 1)).astype(int), values = X, axis = 1) # We are building our numpy array

X_opt = X[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]] # Now we combine all 16 input variables into our first iteration
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() # Now we set up our regressor function with OLS again as before.
print(regressor_OLS.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.024
Model:                            OLS   Adj. R-squared:                  0.024
Method:                 Least Squares   F-statistic:                     452.7
Date:                Fri, 11 Feb 2022   Prob (F-statistic):               0.00
Time:                        17:15:04   Log-Likelihood:            -3.0178e+05
No. Observations:              261642   AIC:                         6.036e+05
Df Residuals:                  261627   BIC:                         6.037e+05
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.7513      0.156     36.769      0.0

In [29]:
# Removing where p-values > 0.05
X_opt = X[:, [0, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]] # variables having p-values > 0.05 have been removed
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() # Now we set up our regressor function with OLS again as before.
print(regressor_OLS.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.024
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     525.6
Date:                Fri, 11 Feb 2022   Prob (F-statistic):               0.00
Time:                        17:16:08   Log-Likelihood:            -3.0179e+05
No. Observations:              261642   AIC:                         6.036e+05
Df Residuals:                  261629   BIC:                         6.037e+05
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.7702      0.156     36.898      0.0

In [30]:
# Train and Test Spliting and also Model testing
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

X_train, X_test, y_train, y_test = train_test_split(X_opt, y, test_size = 0.2, random_state = 0) 

regressor = LinearRegression()
regressor.fit(X_train, y_train)

print('Model accuracy score:', round(regressor.score(X_train,y_train)*100,2))

Model accuracy score: 2.35


Multiple regression improves the accuracy score of the model compared to a simple linear regression. The accuracy improved from 0.1 % to 2.35 % as it uses all the variables in the dataset.

5. Transform your dependent variable from question 1 into a binary shape. Build and run a logistic regression model and explain in AT LEAST two sentences how your output improves on the simple and multiple regression. Is the logistic model better or not as good?

In [31]:
# The Dependant variable convertion to a binary shape as used in the multiple linear regression.
#Reading in the Olympic history: athletes and results  as olympic and results dataframe
olympic = pd.read_csv('archive/athlete_events.csv')
olympic['Medal'] = pd.factorize(olympic['Medal'])[0]
olympic['Medal'].unique()


array([-1,  0,  1,  2], dtype=int32)

In [36]:
# storing only the factor levels array inplace of the object utilizing pandas factorize to utilize all the dataset variables 

olympic['Sex'] = pd.factorize(olympic['Sex'])[0]
olympic['Name'] = pd.factorize(olympic['Name'])[0]
olympic['Team'] = pd.factorize(olympic['Team'])[0]
olympic['NOC'] = pd.factorize(olympic['NOC'])[0]
olympic['Games'] = pd.factorize(olympic['Games'])[0]
olympic['Season'] = pd.factorize(olympic['Season'])[0]
olympic['City'] = pd.factorize(olympic['City'])[0]
olympic['Sport'] = pd.factorize(olympic['Sport'])[0]
olympic['Event'] = pd.factorize(olympic['Event'])[0]
olympic['Age'] = np.nan_to_num(olympic['Age'])
olympic['Age'] = olympic['Age'].astype('int32')
olympic['Height'] = np.nan_to_num(olympic['Height'])
olympic['Height'] = olympic['Height'].astype('int32')
olympic['Weight'] = np.nan_to_num(olympic['Weight'])
olympic['Weight'] = olympic['Weight'].astype('int32')
olympic.shape

(271116, 15)

In [37]:
# Logistic regression Model.

# Training and Test set
dfTrain = olympic[:202000]
dfTest = olympic[202000:205500]
dfCheck = olympic[205500:]

# Conversion to NumPy arrays
train_y = np.asarray(dfTrain['Medal']) # We store target in train_y, but as numpy array
train_x = np.asarray(dfTrain.drop('Medal',1)) # We remove target from the predictors (because it's the output variable)
test_y = np.asarray(dfTest['Medal']) # See the comments above
test_x = np.asarray(dfTest.drop('Medal',1))

# input normalization.
np.seterr(divide='ignore', invalid='ignore')
means = np.mean(train_x, axis=0)
std = np.std(train_x, axis=0)

train_x = np.nan_to_num(train_x)
train_x = (train_x - means)/std
test_x = (test_x - means)/std

# Building the Model
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(train_x, train_y)

# Testing the Model
test_x = np.nan_to_num(test_x)
accuracy = model.score(test_x, test_y)
print("accuracy = ", accuracy * 100, "%")

accuracy =  84.77142857142857 %


When compared to linear regression, the model accuracy has increased dramatically from 2.35 percent to 84.77142857142857 percent. Logistic regression outperforms linear regression by a wide margin.